### ISSUE TITLE
Don't all private method for inheritance

### ISSUE THREAD

**Issue by Progi1984** [PrestaShop/phpstan-prestashop#34]
Rule :
* If the class is final, allow private method
* Else all private must be protected for inheritance

**Comment by jolelievre** [comment]
We shouldn't let a bot decide this for us, it is our job and expertise to decide the field privacy on a case by case approach
Private and protected fields both have their pros and cons, regardless of the fact that the class is final

Imposing this just on the final keyword is just a lazy and risky approach IMHO


It's a HUGE no go for me! (I just refrained myself from closing the issue right away because I'm polite 😅 )

**Comment by matks** [comment]
I agree with @jolelievre that this is not something that should be enforced by a bot.

If we use phpstan-prestashop to enforce this rule it means PRs that do not comply will be REJECTED and not merged. That's too far for a rule that is not a standard best practice agreed to by everyone

**Comment by Progi1984** [comment]
I disagree. 

Why ? If we have a public or protected method  which a private method, how can we correctly override this one  ?

**Comment by matks** [comment]
Jonathan and I are against enforcing this rule on all the codebase with a bot.

You are talking about a single usecase: when a `public` or `protected` method uses internally a `private` method. But your issue description simply says "never use private unless final".

Also on a bigger topic:
- https://en.wikipedia.org/wiki/Composition_over_inheritance
- https://codeburst.io/inheritance-is-evil-stop-using-it-6c4f1caf5117
- https://medium.com/@vinayganeshh/whats-so-bad-about-inheritance-e37195150f7a

**Comment by Progi1984** [comment]
I understand the problem for you in "never use private unless final".

But can you explain in a inheritable class a case where a private exists ? Because I don't see any case.
